### PR TITLE
refactor(connector): update error handling for Paypal, Checkout, Mollie to include detailed messages

### PR DESCRIPTION
**Pull request by prasunna09**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [x] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Modified error message from dynamic to static messages. Also modified the tests accordingly.

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

Screenshot 2023-05-12 at 3 11 12 PM

Screenshot 2023-05-12 at 3 12 06 PM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/tests/connectors/checkout.rs [hunk 1]
```diff
@@ -313,28 +310,6 @@ async fn should_sync_refund() {
 }
 
 // Cards Negative scenerios
-// Creates a payment with incorrect card number.
-#[serial_test::serial]
-#[actix_web::test]
-async fn should_fail_payment_for_incorrect_card_number() {
```
Comment (SanchithHegde) [inline crates/router/tests/connectors/checkout.rs:319]: Why was this test removed? Would using a card number of valid length (16 digits) help?

File: crates/router/tests/connectors/paypal.rs [hunk 2]
```diff
@@ -428,27 +428,6 @@ async fn should_sync_refund() {
 }
 
 // Cards Negative scenerios
-// Creates a payment with incorrect card number.
-#[actix_web::test]
-async fn should_fail_payment_for_incorrect_card_number() {
```
Comment (SanchithHegde) [inline crates/router/tests/connectors/paypal.rs:433]: Why was this test removed? Would using a card number of valid length (16 digits) help?

File: crates/router/src/connector/paypal.rs [hunk 3]
```diff
@@ -71,7 +71,7 @@ impl Paypal {
             .parse_struct("Paypal ErrorResponse")
             .change_context(errors::ConnectorError::ResponseDeserializationFailed)?;
 
-        let message = match response.details {
+        let error_reason = match response.details {
             Some(mes) => {
```
Comment (SanchithHegde) [inline crates/router/src/connector/paypal.rs:75]: This `Some(...)` arm can be succinctly written as:

```rust
Some(order_errors) => order_errors
    .iter()
    .map(|error| {
        let mut reason = format!("description - {}", error.description);
        if let Some(value) = &error.value {
            reason.push_str(&format!(", value - {value}"));
        }
        if let Some(field) = error
            .field
            .as_ref()
            .and_then(|field| field.split('/').last())
        {
            reason.push_str(&format!(", field - {field}"));
        }
        reason.push(';');
        reason
    })
    .collect::<String>(),
```

File: crates/router/src/connector/paypal.rs [hunk 4]
```diff
@@ -168,7 +168,7 @@ impl ConnectorCommon for Paypal {
...
```
Comment (SanchithHegde) [inline crates/router/src/connector/paypal.rs:172]: The `Some(...)` arm can be succinctly written as:

```rust
Some(error_details) => error_details
    .iter()
    .map(|error| format!("description - {} ; ", error.description))
    .collect::<String>(),
```

File: crates/router/tests/connectors/paypal.rs [hunk 5]
```diff
@@ -468,7 +447,7 @@ async fn should_fail_payment_for_incorrect_cvc() {
         .unwrap();
     assert_eq!(
         response.response.unwrap_err().message,
-        "description - The value of a field does not conform to the expected format., value - 12345, field - security_code;",
+        "Request is not well-formed, syntactically incorrect, or violates schema.",
```
Comment (jagan-jaya) [inline crates/router/tests/connectors/paypal.rs:450]: add assertion for the reason where the security code error is notified

File: crates/router/tests/connectors/paypal.rs [hunk 6]
```diff
@@ -490,7 +469,7 @@ async fn should_fail_payment_for_invalid_exp_month() {
         .unwrap();
     assert_eq!(
         response.response.unwrap_err().message,
-        "description - The value of a field does not conform to the expected format., value - 2025-20, field - expiry;",
+        "Request is not well-formed, syntactically incorrect, or violates schema.",
```
Comment (jagan-jaya) [inline crates/router/tests/connectors/paypal.rs:472]: add assertion also for the reason where the expiry date error is notified

File: crates/router/tests/connectors/paypal.rs [hunk 7]
```diff
@@ -512,7 +491,7 @@ async fn should_fail_payment_for_incorrect_expiry_year() {
         .unwrap();
     assert_eq!(
         response.response.unwrap_err().message,
-        "description - The card is expired., field - expiry;",
+        "The requested action could not be performed, semantically incorrect, or failed business validation.",
```
Comment (jagan-jaya) [inline crates/router/tests/connectors/paypal.rs:494]: add assertion also for the reason where the expiry date error is notified

File: crates/router/tests/connectors/paypal.rs [hunk 8]
```diff
@@ -552,7 +531,7 @@ async fn should_fail_void_payment_for_auto_capture() {
         .expect("Void payment response");
     assert_eq!(
         void_response.response.unwrap_err().message,
-        "description - Authorization has been previously captured and hence cannot be voided. ; "
+        "The requested action could not be performed, semantically incorrect, or failed business validation."
```
Comment (jagan-jaya) [inline crates/router/tests/connectors/paypal.rs:534]: add assertion also for the reason where the void is invalid

File: crates/router/tests/connectors/paypal.rs [hunk 9]
```diff
@@ -577,7 +556,7 @@ async fn should_fail_capture_for_invalid_payment() {
         .unwrap();
     assert_eq!(
         capture_response.response.unwrap_err().message,
-        "description - Specified resource ID does not exist. Please check the resource ID and try again. ; ",
+        "The specified resource does not exist.",
```
Comment (jagan-jaya) [inline crates/router/tests/connectors/paypal.rs:559]: add assertion for reason also

File: crates/router/tests/connectors/paypal.rs [hunk 10]
```diff
@@ -600,7 +579,7 @@ async fn should_fail_for_refund_amount_higher_than_payment_amount() {
         )
         .await
         .unwrap();
-    assert_eq!(&response.response.unwrap_err().message, "description - The refund amount must be less than or equal to the capture amount that has not yet been refunded. ; ");
+    assert_eq!(&response.response.unwrap_err().message, "The requested action could not be performed, semantically incorrect, or failed business validation.");
```
Comment (jagan-jaya) [inline crates/router/tests/connectors/paypal.rs:582]: add assertion for reason also

**Review by jagan-jaya — CHANGES_REQUESTED** [review]

File: crates/router/tests/connectors/checkout.rs (hunk 1 above)
Comment (prasunna09) [reply crates/router/tests/connectors/checkout.rs:319]: should_fail_payment_for_incorrect_card_number is removed for this connectors mentioned above . As we have luhn check in place, we don't need this test . this test will only validate the invalid card number scenario and not unsupported card scenarios as connectors don't throw INVALID_CARD_NUMBER type error. This is discussed with @jagan-jaya @ArjunKarthik

File: crates/router/tests/connectors/paypal.rs (hunk 2 above)
Comment (prasunna09) [reply crates/router/tests/connectors/paypal.rs:433]: should_fail_payment_for_incorrect_card_number is removed for this connectors mentioned above . As we have luhn check in place, we don't need this test . this test will only validate the invalid card number scenario and not unsupported card scenarios as connectors don't throw INVALID_CARD_NUMBER type error. This is discussed with @jagan-jaya @ArjunKarthik

**Review by SanchithHegde — DISMISSED** [review]

File: crates/router/src/connector/checkout.rs [hunk 11]
```diff
@@ -92,17 +92,23 @@ impl ConnectorCommon for Checkout {
...
+                .as_ref()
+                .and_then(|error_codes| error_codes.first().cloned())
                 .unwrap_or_else(|| consts::NO_ERROR_MESSAGE.to_string()),
-            reason: None,
+            reason: Some(
+                response
+                    .error_codes
+                    .unwrap_or_else(|| vec![consts::NO_ERROR_CODE.to_string()])
+                    .join(" & "),
+            ),
```
Comment (jagan-jaya) [inline crates/router/src/connector/checkout.rs:111]:
```suggestion
            reason: response.error_codes.map(|errors| errors.join(" & ")),
```

**Review by jagan-jaya — CHANGES_REQUESTED** [review]

**Review by jagan-jaya — APPROVED** [review]

**Review by ArjunKarthik — APPROVED** [review]

**Pull request merged**